File: problem_solving ( C++ )/Random/permutation.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define endl '\n'
#define fastio() ios::sync_with_stdio(false); cin.tie(nullptr);

// void backtrack(vector<int>& nums, vector<bool>& used, vector<int>& path, vector<vector<int>>& result) {
//     if (path.size() == nums.size()) {
//         result.push_back(path);
//         return;
//     }

//     for (int i = 0; i < nums.size(); ++i) {
//         if (used[i]) continue;
//         if (i > 0 && nums[i] == nums[i - 1] && !used[i - 1])
//             continue;

//         used[i] = true;
//         path.push_back(nums[i]);
//         backtrack(nums, used, path, result);
//         path.pop_back();
//         used[i] = false;
//     }
// }

// vector<vector<int>> permutation(vector<int> nums) {
//     sort(nums.begin(), nums.end());
//     vector<vector<int>> result;
//     vector<int> path;
//     vector<bool> used(nums.size(), false);
//     backtrack(nums, used, path, result);
//     return result;
// }

// int main() {
//     fastio();
//     vector<int> s = {1, 2, 3};
//     vector<vector<int>> perms = permutation(s);

//     for (const auto& p : perms) {
//         for (int num : p)
//             cout << num << " ";
//         cout << "\n";
//     }

//     return 0;
// }


int main() {
    vector<int> a = {1, 2, 3};
    sort(a.begin(), a.end());

    do {
        for (int x : a) cout << x << " ";
        cout << "\n";
    } while (next_permutation(a.begin(), a.end()));
}

File: problem_solving ( python )/Random/permutation.py
# def permuteUnique(arr):
#     if len(arr) == 0:
#             return []
#     if len(arr) == 1:
#         return [arr]
        
#     perm = []
#     seen = set() 
#     for i in range(len(arr)):
#         elem = arr[i]
#         if elem in seen:
#             continue
#         seen.add(elem)
#         rem = arr[:i] + arr[i+1:]
#         for p in permuteUnique(rem): 
#             perm.append([elem] + p)
        
#     return perm
# s = [1,2,3]
# for p in permuteUnique(s):
#     print(p)

# def permutation(nums):
#     nums.sort() 
#     result = []
    
#     def backtrack(path, used):
#         if len(path) == len(nums):
#             result.append(path[:])
#             return
#         for i in range(len(nums)):
#             if used[i]:
#                 continue
#             if i > 0 and nums[i] == nums[i - 1] and not used[i - 1]:
#                 continue 
#             used[i] = True
#             path.append(nums[i])
#             backtrack(path, used)
#             path.pop()
#             used[i] = False
    
#     backtrack([], [False] * len(nums))
#     return result

# s = [1, 2, 3]
# for p in permutation(s):
#     print(p)

# from collections import Counter

# def permutation(nums):
#     nums.sort()
#     result = []
#     def backtrack(path, count):
#         if len(path) == len(nums):
#             result.append(path[:])
#             return
        
#         for num in count:
#             if count[num] > 0:
#                 path.append(num)
#                 count[num] -= 1
#                 backtrack(path, count)
#                 path.pop()
#                 count[num] += 1
        
#     count = Counter(nums)
#     backtrack([], count)
#     return result
        
# s = [1, 2, 3]
# for p in permutation(s):
#     print(p)
